internal/replacements/replace_failed_process_groups.go

```diff
@@ -37,10 +40,30 @@ func getMaxReplacements(cluster *fdbv1beta2.FoundationDBCluster, maxReplacements
 		if processGroupStatus.IsMarkedForRemoval() && !processGroupStatus.IsExcluded() {
 			// Count all removals that are in-flight.
 			removalCount++
+			faultDomains[processGroupStatus.FaultDomain] = fdbv1beta2.None{}
 		}
 	}
 
-	return maxReplacements - removalCount
+	return maxReplacements - removalCount, faultDomains
+}
+
+// removalAllowed will return true if the removal is allowed based on the clusters automatic replacement configuration.
+func removalAllowed(cluster *fdbv1beta2.FoundationDBCluster, maxReplacements int, faultDomainsWithReplacements map[fdbv1beta2.FaultDomain]fdbv1beta2.None, faultDomain fdbv1beta2.FaultDomain) bool {
+	if cluster.FaultDomainBasedReplacements() {
+		_, faultDomainHasReplacements := faultDomainsWithReplacements[faultDomain]
+		// The fault domain has already a replacement ongoing, so we are able to approve the replacement of the failed
+		// process group.
+		if faultDomainHasReplacements {
+			return true
```

Comment:
Still learning obviously, but this *feels* worrisome to me as we could end up exceeding the max concurrent replacements to a much greater degree when we allow all the processes in it go down.  
In my head I'm imagining a scenario where one processGroup in each fault domain has problems, is marked for replacement, but that leads to all fault domains getting an OK to replace all processGroups if that's somehow possible?

Reply:
ok that makes sense